These changes remove PCAST errors within the calcksmax subroutine.

sherimickelson opened the pull request:
Adding these changes removed the PCAST errors from the calcksmax subroutine.

sherimickelson committed:
These changes remove PCAST errors within the calcksmax subroutine.

johnmauff reviewed (commented):
@sherimickelson, this appears to use the old version of calccfl instead of the newly NVIDIA provided calccfl_fold version.  Sorry that I did not mentioned this previous.

sjsprecious commented on src/misclibs.F:
```diff
@@ -3482,8 +3482,9 @@ subroutine calcksmax(nstat,rstat,dt,uh,vh,mf,kmh,kmv,khh,khv)
 
-!$acc data copyout (ksh,ksv,imaxth,jmaxth,kmaxth,imaxtv,jmaxtv,kmaxtv) &
-!$acc      copyin (khh,vh,uh,mf)
+!!$acc data copyout (ksh,ksv,imaxth,jmaxth,kmaxth,imaxtv,jmaxtv,kmaxtv) &
+!!$acc      copyin (khh,vh,uh,mf)
```
remove these two lines if they are wrong.

sherimickelson replied:
Thanks @sjsprecious , I removed these lines

sjsprecious commented on src/misclibs.F:
```diff
@@ -3523,20 +3524,29 @@ subroutine calcksmax(nstat,rstat,dt,uh,vh,mf,kmh,kmv,khh,khv)
             do i=1,ni
                tem = khh(i,j,k)*dtdx*uh(i)*uh(i)
                if( tem.eq.ksh(k) )then
+                 !$acc atomic write
```
I am still confused about the behavior of `acc atomic write`. I would suggest using `loop seq` for the nested three loops to behave the same as the serial code first.

sherimickelson replied:
@sjsprecious , I'm going to leave those lines in for the moment because as @johnmauff points out, this might not be an issue because the NVIDIA version does not use these.  Once we've verified the correctness of the new version, we can remove this code.

sjsprecious commented on src/misclibs.F:
```diff
@@ -3551,18 +3561,26 @@ subroutine calcksmax(nstat,rstat,dt,uh,vh,mf,kmh,kmv,khh,khv)
       jmaxv=1
       kmaxv=1
-      !$acc parallel loop gang vector default(present)
+      !$acc parallel loop default(present)
```
Change it to `loop seq` for the same reason above.

sherimickelson replied:
Thanks @sjsprecious , I made this change.

sherimickelson committed:
Create a new version of calcksmax (calcksmax_fold) that uses new method created by NVIDIA

sherimickelson commented:
@sjsprecious and @johnmauff I created a new version of calcksmax and called it calcksmax_fold.  statpack.F calls this new version.  It doesn't seem to have changed any of the stat variable output and this still removed the PCAST error.

sjsprecious commented:
Hi @sherimickelson , thanks for this PR. I checked out the current `gpu-opt` branch and performed a CPU run. Then I checked out the `calcksmax_fix` branch and performed an MPI+GPU run. Both runs turned off the `tqnudge`. My results indicated that there were lot of differences in the stat output like the variables `KSHMAX` and `KSVMAX`.

sherimickelson commented:
@sjsprecious , thanks for checking for this.  My previous comment was in regards of the answers not being changed between [91846fb](https://github.com/george-bryan/CM1/pull/53/commits/91846fbfdbac673328615bdc09afbe6489070622) and [48d9d0e](https://github.com/george-bryan/CM1/pull/53/commits/48d9d0e035897642aa306d2dacf0fa183312643d).  

In regards to the answers being different between the CPU and GPU version, this is the tougher question to answer :).   I would expect that this changed answers because of the PCAST error that was coming out of it is now fixed and those are the two variables I modified.  Do you see anything incorrect in the code that would indicate that further modifications are needed?

sjsprecious commented on src/misclibs.F:
```diff
@@ -3630,6 +3644,210 @@ subroutine calcksmax(nstat,rstat,dt,uh,vh,mf,kmh,kmv,khh,khv)
+      RANGE=(ni*nj+1)/2       !! Round up.
+      do while ( RANGE > 1 )
+!! Now fold the 2-d array with half the threads each time.
+!$acc parallel loop gang vector private(i)
```
according to the comments, should `acc parallel loop seq` be used here?

sjsprecious commented on src/misclibs.F:
```diff
@@ -3630,6 +3644,210 @@ subroutine calcksmax(nstat,rstat,dt,uh,vh,mf,kmh,kmv,khh,khv)
+        PAIRS=RANGE/2
+        RANGE=(RANGE+1)/2
+      enddo 
+
```
I think an `acc update host (ksh,imaxth,jmaxth,kmaxth,ksv,imaxtv,jmaxtv,kmaxtv)` directive is needed here.

sherimickelson replied:
Thanks @sjsprecious for catching this.  I believe you're correct and I'll add the update host and verify correctness.

sherimickelson replied:
calccfl_fold will need this change as well

sherimickelson replied:
Both of these updates to the host were added in [e00ae25](https://github.com/george-bryan/CM1/pull/53/commits/e00ae254d365a73a0f2eaee2381c989be5fd04ff)

sherimickelson committed:
Add an update host to calcksmax_fold and calccfl_fold

sherimickelson commented:
This latest commit ([e00ae25](https://github.com/george-bryan/CM1/pull/53/commits/e00ae254d365a73a0f2eaee2381c989be5fd04ff)) results in
26 stat variables are identical.
62 stat variables show a mean relative difference > 1e-06
0 stat variables show a mean relative difference <= 1e-06
Top 4 variables are THPMAX, STHPMX, STHPMN, TMW

sjsprecious commented:
Thanks @sherimickelson for the update. The stat output looks worse compared to the current `gpu-opt` branch. Have you already merged the latest commit (https://github.com/george-bryan/CM1/commit/fd054dfb6648b9fe833df2d0435f12dca8c96044) in `gpu-opt` to your branch?

sherimickelson commented:
No, those stat numbers are without the merge.

sherimickelson commented:
@sjsprecious , with the merge, I'm now seeing
28 stat variables are identical.
54 stat variables show a mean relative difference > 1e-06
6 stat variables show a mean relative difference <= 1e-06

sjsprecious commented:
> @sjsprecious , with the merge, I'm now seeing 28 stat variables are identical. 54 stat variables show a mean relative difference > 1e-06 6 stat variables show a mean relative difference <= 1e-06

Thanks @sherimickelson. The new results show fewer variables that are either identical or larger than the tolerance but more variables that are smaller than the tolerance. To me this is fine unless we want to strictly keep the number of identical variables. If you have pushed all your changes to the `calcksmax_fix` branch, I would like to run it on my end and double check the results.

sherimickelson commented:
Thanks, @sjsprecious .  Yes, I've pushed all of my changes and that would be great if you can double check my results.

sjsprecious commented:
My results for verification runs:
CPU vs. OpenACC:
Identical variables: 30
variables with red_diff > 1.e-6: 53
variables with red_diff <= 1.e-6: 5
four variables with largest error: KSVMAX, KSHMAX, TMW, TKEMAX

CPU vs. MPI+OpenACC:
Identical variables: 27
variables with red_diff > 1.e-6: 58
variables with red_diff <= 1.e-6: 3
four variables with largest error: KSVMAX, KSHMAX, DIVMIN, TKEMAX
